hash_practice/exercises.py

```diff
@@ -1,19 +1,50 @@
+from collections import defaultdict
 
 def grouped_anagrams(strings):
     """ This method will return an array of arrays.
         Each subarray will have strings which are anagrams of each other
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n^2) rip
+        Space Complexity: O(n)
     """
-    pass
+    res = set()
+    freq_maps = {}
+    for word in strings:
+        freq_map = {}
+        for char in word:
+            if freq_map.get(char):
+                freq_map[char] += 1
+            else:
+                freq_map[char] = 1
+        freq_maps[word] = freq_map
+    for word in strings:
+        sub_res = set()
+        for key, map in freq_maps.items():
+            if freq_maps[word] == map:
+                sub_res.add(key)
+        res.add(tuple(sub_res))

```

Comment:
This definitely works, but as you note, this has O(n^2) time.

I do gather from the solutions to the other problems in this set, you do have a good handle, but you can get this down to O(n) by more directly using a dictionary as a hash table.

That is, rather than first making a dictionary mapping words to frequency maps and then going through that to determine which words share frequency maps, could you build a dictionary that maps the frequency maps to the list of words that have them? If you can do that you can get this down to O(n).

(You'd also get less guff from the tests for the order not matching, but that's less important.)

As reluctant as I am to do it due to your good work on the other parts, I'm going to mark this as Yellow for no other reason than to encourage you to try solving it in O(n) as described.